docs/lang.md

```diff
@@ -259,13 +259,176 @@ The macroexpand-1 function is just like macroexpand except it will not
+returned.
+
+### let vs let\*
+
+`let` and `let*` are used to create bindings for local variables within a 
+new scope.  `let` bindings cannot refer to each other. `let*` bindings can
+refer to each other (left-to-right, top-to-bottom). The result of the
+evaluation of the last expression within the `let` is returned.
+
+```lisp
+(let ((variable1 result1)
+      (variable2 result2) ; result2 cannot reference variable1
+      ...
+      (variable3 result3))
+  expr1 ; some expression that can use bound variables
+  ...
+  exprN ; exprN evaluation is returned
+  )
+```
+
+```lisp
+(set 'x 0)
+(let ([x (+ x 1)]
+      [x (+ x 1)])
+  x)                ; evaluates to 1
+```
+
+```lisp
+(set 'x 0)
+(let* ([x (+ x 1)]
+       [x (+ x 1)])
+  x)                ; evaluates to 2
+```
+
+### flet vs labels
+
+`flet` and `labels` are used to create bindings for local functions within a
+new scope.  `flet` bindings are not recursive and cannot refer to each other.
+`labels` bindings can be recursive and can refer to each other (left-to-right,
+top-to-bottom).
+
+```lisp
+(flet ((func1 (arg1 ... argn) expr1 ... exprN)
+       ...
+       (func2 (arg1 ... argn) expr1 ... exprN))
+  expr1 ; some expression that can use bound functions
+  ...
+  exprN
+  ) ; exprN evaluated and returned
+```
+
+```lisp
+(defun count () 0)
+(flet ([count () (+ (count) 1)]
+       [count () (+ (count) 1)])
+  (count)) ; evaluates to 1
+```
+
+```lisp
+(defun count () 0)
+(labels ([count0 () (+ (count) 1)]
+	 [count1 () (+ (count0) 1)])
+  (count1)) ;  to 2
+```
+
+### macrolet
+
+`macrolet` is used to create bindings for local macros within a new scope, in
+an analogous way as `flet` and `labels`.
+
+### assert
+
+`assert` takes an expression and optional string, and evalutes the expression.
+If the result of the evaluation is truthy then assert returns `()`, otherwise
+`assert` will output the assertion failure message to stderr and raise an
+error.
+
+### progn
+
+`progn` causes each of its arguments to be evaluated in sequence and then
+returns the value of the last one. The preceding expressions are evaluated
+only for the side effects they perform. The values produced by them are
+discarded.
+
+```lisp
+(progn
+  expr1
+  expr2
+  ...
+  exprN) ; exprN evaluation is returned
+```
+
+### thread-first, thread-last
+
+`thread-first` and `thread-last` help make nested function calls more readable,
+and function similar to the clojure `->` and `->>` macros.
+
+The word "thread" in this context (meaning passing a value through a pipeline
+of functions) is unrelated to the concept of concurrent threads of execution.
+
+`thread-first` takes the first argument and passes it as an argument to the
+function defined in the second argument. The result of evaluating this
+expression is then passed to the function defined in the third argument, and
+so on.
+
+```lisp
+(defun add1 (x) (+ x 1))
+(add1 (add1 2))                ; evaluates to 4
+(thread-first 1 (add1) (add1)) ; evaluates to 4
```

Comment:
This example is wrong. I guess you probably wanted the first argument of `thread-first` to be 2 so that it matches the previous line.